File: MMORPGServer/Proto/CreateLuaProto/RoleOperation_SelectRoleInfoProto.lua
--客户端查询角色信息
RoleOperation_SelectRoleInfoProto = { ProtoCode = 10009, RoldId = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleOperation_SelectRoleInfoProto.__index = RoleOperation_SelectRoleInfoProto;

function RoleOperation_SelectRoleInfoProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleOperation_SelectRoleInfoProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleOperation_SelectRoleInfoProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.RoldId);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleOperation_SelectRoleInfoProto.GetProto(buffer)

    local proto = RoleOperation_SelectRoleInfoProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.RoldId = ms:ReadInt();

    ms:Dispose();
    return proto;
end

File: MMORPGServer/GameServerApp/GameServerApp/Proto/Role/RoleOperation_SelectRoleInfoProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2020-03-15 10:58:27
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 客户端查询角色信息
/// </summary>
public struct RoleOperation_SelectRoleInfoProto : IProto
{
    public ushort ProtoCode { get { return 10009; } }

    public int RoldId; //角色编号

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(RoldId);
            return ms.ToArray();
        }
    }

    public static RoleOperation_SelectRoleInfoProto GetProto(byte[] buffer)
    {
        RoleOperation_SelectRoleInfoProto proto = new RoleOperation_SelectRoleInfoProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.RoldId = ms.ReadInt();
        }
        return proto;
    }
}